analysis_functions.py

```diff
@@ -0,0 +1,160 @@
+    
+    grouped_ts = df_timeseries.group_by('channel_id').mean()[['channel_id',feature_to_divide_by]]
+    merged = df_vid.join(grouped_ts, on='channel_id')
+    return merged.with_columns(pl.col(['dislike_count','duration','like_count','view_count','num_comms']) / merged[feature_to_divide_by])
+
+def ttest_between_events (statistic_event1, statistic_event2):
+    event1 = statistic_event1
+    event2 = statistic_event2
+    event1 = event1.drop_nulls()
+    event2 = event2.drop_nulls()
+    return stats.ttest_ind(event1,event2)
+
+def Ftest_between_events (statistic_event1, statistic_event2):
+    event1 = statistic_event1
+    event2 = statistic_event2
+    event1 = event1.drop_nulls()
+    event2 = event2.drop_nulls()
+    return stats.f_oneway(event1,event2)
+
+def get_general_vid_statistics(filtered_df, cols_to_keep = ['dislike_count','duration','like_count','view_count']):
+
+    '''get general statistics for all channel on a given column'''
+
+    means = filtered_df.mean()[cols_to_keep]
+    stdevs = filtered_df.std()[cols_to_keep]
+    medians = filtered_df.median()[cols_to_keep]
+    
+    return means,stdevs,medians
+
+def compare_overall_vid_count_between_events (vid_count_1,vid_count_2):
+    
+    print ('Total number of videos for event 1 : ', vid_count_1['counts'].sum())
+    print ('Total number of videos for event 2 : ', vid_count_2['counts'].sum())
+    print ('Difference in videos from event 1 to event 2 : ', vid_count_1['counts'].sum() - vid_count_2['counts'].sum())
+    return vid_count_1.join(vid_count_2, on='channel_id').rename({'counts':'counts_1', 'counts_right':'counts_2'})
+
+def compare_video_statistics_between_events (videos_1, videos_2):
+    
+    v_means_1,v_stdevs_1,v_medians_1 = get_general_vid_statistics(videos_1)
+    v_means_2,v_stdevs_2,v_medians_2 = get_general_vid_statistics(videos_2)
+    
+    print ('Average number of views for each video in event 1 : ', v_means_1['view_count'].sum())
+    print ('Average number of views for each video in event 2 : ', v_means_2['view_count'].sum())
+    print ('Difference average number of views for each video from event 1 to event 2 : ', v_means_1['view_count'].sum() - v_means_2['view_count'].sum())
+    return pl.concat([v_means_1,v_means_2]).insert_column(0,pl.Series(['event_1','event_2']))
+
+def plot_covariance (df, titlecov, titlehist):
+    cov = np.cov(df.drop('channel_id').drop_nulls())
+    plt.imshow(cov)
+    plt.colorbar()
+    plt.title(titlecov)
+    plt.tight_layout()
+    plt.show()
+
+    #plot histogram of correlations to know how to evaluate the data.
+    plt.hist(cov.flatten())
+    plt.yscale('log')
+    plt.title(titlehist)
+    plt.ylabel('Log count')
+    plt.xlabel('Covariance')
+    plt.show()
+    
+    return cov
+
+def get_correlated_channels (df, threshold):
+    cov = np.cov(df.drop('channel_id').drop_nulls()) 
+    
+    first_correlated_channel = df[np.where(np.tri(len(cov), k=-1) * cov > threshold)[0]]
```

Comment:
This computes the covariance between all the channels? And then you get two random channels which have a correlation above a given threshold?